File: sdks/python/greplog/ulid.py
"""ULID generation matching the Node SDK implementation."""

from __future__ import annotations

import random

_ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
_ENCODING_LEN = len(_ENCODING)

_last_timestamp = 0
_last_random = 0


def generate_ulid() -> str:
    global _last_timestamp, _last_random

    now = int(__import__("time").time() * 1000)
    if now != _last_timestamp:
        _last_timestamp = now
        _last_random = 0
    _last_random += 1

    return _timestamp_to_ulid(now) + _random_to_ulid(_last_random)


def _timestamp_to_ulid(ts: int) -> str:
    chars: list[str] = []
    val = ts
    for _ in range(10):
        chars.append(_ENCODING[val % _ENCODING_LEN])
        val //= _ENCODING_LEN
    return "".join(reversed(chars))


def _random_to_ulid(seq: int) -> str:
    chars: list[str] = []
    for i in range(16):
        if i < 4:
            idx = random.randint(0, _ENCODING_LEN - 1)
        else:
            idx = ((seq >> ((i - 4) * 5)) + random.randint(0, 7)) % _ENCODING_LEN
        chars.append(_ENCODING[idx])
    return "".join(chars)

File: sdks/node/src/ulid.ts
const ENCODING = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
const ENCODING_LEN = ENCODING.length;

let lastTimestamp = 0;
let lastRandom = 0;

function randomChar(): string {
  return ENCODING[Math.floor(Math.random() * ENCODING_LEN)];
}

export function generateULID(): string {
  const now = Date.now();
  if (now !== lastTimestamp) {
    lastTimestamp = now;
    lastRandom = 0;
  }
  lastRandom++;

  const timestampStr = timestampToULID(now);
  const randomStr = randomToULID(lastRandom);

  return timestampStr + randomStr;
}

function timestampToULID(ts: number): string {
  let str = '';
  let val = ts;
  for (let i = 0; i < 10; i++) {
    str = ENCODING[val % ENCODING_LEN] + str;
    val = Math.floor(val / ENCODING_LEN);
  }
  return str;
}

function randomToULID(seq: number): string {
  let str = '';
  for (let i = 0; i < 16; i++) {
    const idx = i < 4
      ? Math.floor(Math.random() * ENCODING_LEN)
      : ((seq >> ((i - 4) * 5)) + Math.floor(Math.random() * 8)) % ENCODING_LEN;
    str += ENCODING[idx];
  }
  return str;
}
